README.md
# NFT Buttholes

Cause butts.

General:
- anyone 18+ can participate
- easy to use, not convoluted; no hidden information
- fair
- NFTs (of buttholes)

Operation:
- EVM compatible smart contract
- can be upgraded
- fair distribution
- no batch minting; no cumulative benefits
- QoL scripts

Trading:
- auctionable
- royalties 

Utility:
- ongoing VIP membership benefits for sites, events, etc

---

### Release Checklist

| Feature | D.evelopment | Staging | Production |
|:-------:|:-----------:|:-------:|:----------:|
| EVM smart contract | :ballot_box_with_check: | :ballot_box_with_check: | :white_large_square: |
| Butthole NFTs | :ballot_box_with_check: | :ballot_box_with_check: | :white_large_square: |
| "VIP" Benefits | :ballot_box_with_check: | :white_large_square: | :white_large_square: |
| 18+ Age Check | :ballot_box_with_check: | :ballot_box_with_check: | :white_large_square: |
| A pic of my Butthole | :ballot_box_with_check: | :white_large_square: | :white_large_square: |
| ...? | :white_large_square: | :white_large_square: | :white_large_square: |


CHANGELOG.md
# Changelog

**0.0.0 : 4/22/2022**
	- setup
	- "majority" of contract
	- tests setup
	**0.0.1 : 4/23/2022**
	- added PaymentSplitPusher
	- added default uri for public "viewing"
	- PaymentSplitPusher skele
	- tests for PaymentSplitPusher
	- added VIP membership "utility"
	-- [utility] write javascript code that can be used to check balance for NFT ownership
	**0.0.2 : 4/24/2022**
	- added "starving artists" address fund / finished PaymentSplitPusher
	-- created modified paymentsplitter that pushes instead of waits for pulls
	- compiled contracts
	- added PaymentSplitPusher to Butthole's royalties process
	**0.0.3 : 4/25/2022**
	- test minting
	- test "butthole" uri uploads
	- Butthole.sol tests complete
	- tested PaymentSplitPusher
	- began writeup for NFT steps
	**0.0.4 : 4/26/2022**
	- updated PaymentSplitPusher interaction to ensure royalties are split to butthole artists before / with donors
	- added fix for paymentsplitter royalties
	- added main.js for bundling
	- renamed PaymentSplitPusher to CheekSpreader... cause duh
	- created single page w/ 18+ button for viewing to ensure actual images are age gated; requires confirming via adding minter role
	- continued write up for project "explainer"
	**0.0.5 : 4/27/2022**
	- finished tests for Buttholes.sol (again)
	**0.0.6 : 5/1/2022**
	- more debugging main.js & butthole.js
	- figured out how metadata.json fits
	**0.0.7 : 5/2/2022**
	- finished `butthole.js` script in js/
	- decided on method for inputting new butthole data to butthole.js args -> {} in a file? runtime args? ----> solution: runtime args
	- finished ipfs metadata functions
	- added extra CheekSpreader function that can be accessed only once by owner of contract for updating new buttholes -> updateCheekSpreader
	- full(ish) implementation of ipfs & html
	**0.0.8 : 5/3/2022**
	- add test for updateCheekSpreader function
	- add test for renounceButthole function
	- test butthole.js
	- documented butthole.js
	**0.0.9 : 5/4/2022**
	- finished testing butthole.js contract interactions
	- continued debugging butthole.js ipfs stuff
	- added ipfs directory setups / checks
	- data writes to ipfs successfully -> requires updating the naming scheme to prevent overwriting issues
	**0.0.10 : 5/9/2022**
	- updated butthole.js w/ command line help menu; restructured args & options
	- [writeup] steps for minting / uploading / creating an nft
	-- accessible method for updating / uploading butthole pics
	-- accessible method for minting
	-- instructions for how to send me info to add a butthole & how I do so
	**0.0.11 : 5/10/2022**
	- documented main.js
	- finished debugging web3.0 interactions in main.js; connects to web3.0 & contract
	- tested main.js
	- tested contract web interface buttons
	- updates to index.html
	**0.0.12 : 5/11/2022**
	- restructured js scripts
	- added 'process' file shim @ project root for browserify
	**0.0.13 : 5/12/2022**
	- added test file for js/butthole.js
	- added test file for js/main.js
	- added tests for birthday date string & eth address validation
	- almost completed test/butthole.js tests
	**0.0.14 : 5/13/2022**
	- mostly finished testing test/test_butthole.js
	- created testing file for main.js
	- finished IPFS metadata test in test_butthole.js
	**0.0.15 : 6/17/2022**
	- updated / fixed metadata format --> 'properties'
	- created schema for metadata
	**0.0.16 : 6/21/2022**
	- updated contract to make adding a butthole w/ or w/o donors simpler
	- updated contract for fetching butthole uri of owners
	- updated minty addon interactions for cli
	**0.0.17 : 7/22/2022**
	- updated Buttholes.js comments to match comment style found in Minty
	- updated cli.js to match Minty's minty.js style
	- combined ipfs.js with Minty's minty.js ipfs code into "minty butthole addon" -> remove current ipfs
	- updated / tested basic minty addon interaction
**0.0.18 : 11-27-2023**
- project revisit / cleanup
- moved excess parts to /notes/old
**ARCHIVING : 11-28-2023**
- end of project --> do something else instead of only buttholes
- cleaned up .env history
- finish long-term plan for how this project is going to work while shooting scenes (hint: it's not)

------------------------------------------------------------------------

# TODO (ARCHIVED)

(cleanup)
- add git branches for staging & production
- finish writing & testing multiple butthole contract copies
- reorg js/ into actual scripts
- separate web component from js/
- finish removing ipfs
- clean package.json (if necessary)

(smart contract)
- rewrite main smart contract to be super simple ERC-721 --> 1 deployed per butthole
- include PaymentSplitter in royalty process
- store butthole jpeg on blockchain
- add ETH purchase cost to minting
- test added ETH costs for minting
- write new tests for new Butthole smart contract
- write / update tests for PaymentSplitter 

(interface)
- add drizzle web interface
- finish testing main.js w/ added drizzle interface

(testing)
- test all changes made to javascript cli & contract interface
- completely finish all major changes and rerun tests for test_butthole.js
- finish testing contract web interface buttons (after drizzle update)
- write tests for updated smart contracts

(staging)
- update instructions for performers / butthole artists 
- update instructions for myself for later

(production)
- add final address of contract on mainnet to etherscan links in docs
- verify smart contract on etherscan

(loopring)
is this even possible? or helpful?
- implement Loopring counterfactual nfts & submitMintNft functionality
- add tests for Loopring interactions

# Dev

requires node @ 16.0.0

http-server

ganache-cli -m "cloud sting village source peace dinner dance fringe slice mandate lens upon" -i 1337

Available Accounts
==================
(0) 0xbC6bC940487C13208FDD7f02d5eF9d9FC886700f (1000 ETH)
(1) 0xA17A41B08a536b004C3Ff379298bAeadE98171B3 (1000 ETH)
(2) 0xc9Ff2b62CF14941d96bAAFecFEBae1a83fd82a7F (1000 ETH)
(3) 0x3Ac74B86550a493EA3F1D8Da7685AeD17CD7a524 (1000 ETH)

js/index.js
const commander = require('commander'),
	  Command = commander.Command;
const path = require('path');

const { add, confirm, getButthole, mint, update, renounce } = require('./butthole.js');

// TODO
// this will totally not work
// figure out how to require the helpers correctly or move them to minty
const { MakeMinty } = require('minty-fresh');
const CONTRACT_NAME = "Buttholes";

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
 * @dev Add an artist, confirm, mint, update starving artists, or renounce a butthole NFT.
 */
async function main() {

	const program = new Command();

	program
	  .name('Minty Buttholes')
	  .description('CLI to some JavaScript NFT utilities')
	  .version('0.0.13');

	program.command('add')
	  .description('Add a butthole template of a performer for minting. Caller must be contract owner.')
            .description('add an NFT schema to IPFS as an available template')
	  .argument('<address>', 'The butthole artist\'s ETH address', parseAddress)
	  .argument('<image>', 'The local path to the butthole image')
	  .option('-n, --name <string>', 'The artist\'s name')
	  .option('-d, --description <string>', 'The artist\'s description')
	  .option('-b, --birthday <date>', 'The artist\'s birthday (MM/DD/YYYY)', parseDate)
	  .option('-s, --starve [addresses...]', 'Up to 3 starving artist ETH addresses', 'specify at least 1 starving artist')
	  .addHelpText('after', `
Example call:
 $ butthole add 0x00.. /path/to/image.jpg -n "My Name" -d "A description." -b "06/06/1990 -s 0x01.. 0x02.. 0x03..`)
	  .action(addButthole);

	program.command('confirm')
	  .description('Add caller as a minter to confirm 18+ age requirement.')
	  .action(confirm);

	program.command('mint')
	  .description('Mint a butthole.')
	  .option('-a, --address <address>', 'An optional receiving address. Defaults to caller', parseAddress)
	  .option('-b, --butthole <id>', 'Specify the butthole to mint by id', parseId)
	  .option('-r, --random', 'Mint a random butthole (default)')
	  .addHelpText('after', `
Example call:
 $ butthole mint -a 0x00.. -r`)
	  .action(mintButthole);

	program.command('update')
	  .description('Update your starving artist(s).')
	  .option('-a, --address <address>', 'The address of the artist to update. Defaults to caller', parseAddress)
	  .requiredOption('-s, --starve [addresses...]', 'Up to 3 \"starving artist\" ETH addresses', 'specify at least 1 starving artist')
	  .addHelpText('after', `
Example call:
 $ butthole update -a 0x00.. -s 0x01.. 0x02.. 0x03..`)
	  .action(async (artist, options) => {
	  	await update({
	  		'artist': artist,
			'starvingArtists': options.starve
	  	});
	  });

	program.command('renounce')
	  .description('Renounce your butthole. You can only renounce your own butthole.')
	  .action(renounce);

	const rootDir = path.join(__dirname, '..');
	process.chdir(rootDir);

	return program;
}
module.exports = main;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

// ---- command action functions

async function addButthole(artist, image, options) {
    const minty = await MakeMinty(CONTRACT_NAME);
    const {assetCid, assetURI, assetGatewayURL} = await minty.uploadAssetData(image, "buttholes");
    options.artist = artist;
    options.image = assetCid;
  	const schema = await selectSchema("butthole");
	const butthole = await promptNFTMetadata(schema, options);
    await add(butthole);
    console.log('🌿 Added a new butthole: ');
    alignOutput(
        ['Contract Name:', chalk.green(minty.name)],
        ['Contract Address:', chalk.yellow(minty.contract.address)],
        ['Butthole Artist:', chalk.green(artist)],
        ['Asset Address:', chalk.blue(assetURI)],
        ['Asset Gateway URL:', chalk.blue(assetGatewayURL)]);
    console.log(colorize(JSON.stringify(butthole), colorizeOptions));
}

async function mintButthole(options) {
    const minty = await MakeMinty(CONTRACT_NAME);
    options.schema = "butthole";
    options.skipMint = true;
    if (options.random) options.butthole = -1;
    options.butthole = await getButthole(options.butthole);
    const nft = await minty.createNFT(options);
    await mint(options.address, nft.metadataURI);
    console.log('🌿 Minted a new butthole: ');
    alignOutput(
        ['Contract Name:', chalk.green(minty.name)],
        ['Contract Address:', chalk.yellow(minty.contract.address)],
        ['Token ID:', chalk.green(nft.tokenId)],
        ['Metadata Address:', chalk.blue(nft.metadataURI)],
        ['Metadata Gateway URL:', chalk.blue(nft.metadataGatewayURL)],
        ['Asset Address:', chalk.blue(nft.assetURI)],
        ['Asset Gateway URL:', chalk.blue(nft.assetGatewayURL)]);
    console.log(colorize(JSON.stringify(nft.metadata), colorizeOptions));
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

// Helpers //

// ensure value is an 0x address
function parseAddress(value) {
	if (!ethers.utils.isAddress(value))
	    throw new commander.InvalidArgumentError('Not an ETH address.');
	return value;
}

function parseId(value) {
	return value;
}

// ensure value is a date value that equates to a unix timestamp
function parseDate(value) {
	if (!isValidDate(value))
	    throw new commander.InvalidArgumentError('Not a date.');
	return value;
}

// https://stackoverflow.com/questions/6177975/how-to-validate-date-with-format-mm-dd-yyyy-in-javascript
function isValidDate(dateString) {
    // First check for the pattern
    if(!/^\d{1,2}\/\d{1,2}\/\d{4}$/.test(dateString))
        return false;

    // Parse the date parts to integers
    var parts = dateString.split("/");
    var day = parseInt(parts[1], 10);
    var month = parseInt(parts[0], 10);
    var year = parseInt(parts[2], 10);

    // Check the ranges of month and year
    if(year < 1000 || year > 3000 || month == 0 || month > 12)
        return false;

    var monthLength = [ 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 ];

    // Adjust for leap years
    if(year % 400 == 0 || (year % 100 != 0 && year % 4 == 0))
        monthLength[1] = 29;

    // Check the range of the day
    return day > 0 && day <= monthLength[month - 1];
}

function alignOutput(labelValuePairs) {
    const maxLabelLength = labelValuePairs
      .map(([l, _]) => l.length)
      .reduce((len, max) => len > max ? len : max);
    for (const [label, value] of labelValuePairs) {
        console.log(label.padEnd(maxLabelLength+1), value);
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

if (require.main === module)
	main().then(async (program) => {
		await program.parseAsync(process.argv)
	    process.exit(0);
	}).catch(err => {
	    console.error(err);
	    process.exit(1);
	})
